docs/ai/reference.rst

```diff
@@ -0,0 +1,357 @@
+        {'Mars Skies on Mars are red.'}
+
+
+------------
+
+
+.. eql:function:: ext::ai::search( \
+                    object: anyobject, \
+                    query: <array>float32 \
+                  ) -> optional tuple<object: anyobject, distance: float64>
+
+    Search an object using its :ref:`ai::index <ref_guide_ai_reference_index>`
+    index.
+
+    Returns objects that match the specified semantic query and the
+    similarity score.
+
+    .. note::
+
+        The ``query`` argument should *not* be a textual query but the
+        embeddings generated *from* a textual query. To have EdgeDB generate
+        the query for you along with a text response, try :ref:`our built-in
+        RAG <ref_ai_overview_rag>`.
+
+    .. code-block:: edgeql-repl
+
+        db> with query := <array<float32>><json>$query
+        ...   select ext::ai::search(Knowledge, query);
+        {
+          (
+            object := default::Knowledge {id: 9af0d0e8-0880-11ef-9b6b-4335855251c4},
+            distance := 0.20410746335983276
+          ),
+          (
+            object := default::Knowledge {id: eeacf638-07f6-11ef-b9e9-57078acfce39},
+            distance := 0.7843298847773637
+          ),
+          (
+            object := default::Knowledge {id: f70863c6-07f6-11ef-b9e9-3708318e69ee},
+            distance := 0.8560434728860855
+          ),
+        }
+
+
+HTTP endpoints
+==============
+
+
+RAG
+---
+
+``https://<edgedb-host>:<port>/branch/<branch-name>/ai/rag``
+
+Responds with text generated by the specified text generation model.
+
+
+Request body
+^^^^^^^^^^^^
+
+* ``model`` (string, required): The name of the text generation model to use.
+
+* ``query`` (string, required): The query string use as the basis for text
+  generation.
+
+* ``context`` (object, required): Settings that define the context of the
+  query.
+
+  * ``query`` (string, required): Specifies the query to determine the relevant
+    objects and index to serve as context for text generation.
+
+  * ``variables`` (object, optional): A dictionary of variables for use in the
+    context query.
+
+  * ``globals`` (object, optional): A dictionary of globals for use in the
+    context query.
+
+  * ``max_object_count`` (int, optional): Maximum number of objects to return;
+    default is 5.
+
+* ``stream`` (boolean, optional): Specifies whether the response should be
+  streamed. Defaults to false.
+
+* ``prompt`` (object, optional): Settings that define a prompt, overriding the
+  default prompt.
+
+  * ``name`` (string, optional): The name of the custom prompt.
+
+  * ``id`` (string, optional): A unique identifier for the custom prompt.
+
+  * ``custom`` (array of objects, optional): Custom prompt messages, each
+    containing a ``role`` and ``content``.
+
+
+Response
+^^^^^^^^
+
+A JSON object or a stream of events, depending on the request body's ``stream``
+parameter, containing the AI-generated data or error messages if the request
+failed.
```

Comment:
Yeah, we probably want more precision in the return type here as well. What is the shape of the JSON data that is returned? What does the stream data look like? Maybe some reference to [server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)  might be useful?

Reply:
I provided examples of both successful and error responses from each of the endpoints, and I better explained the SSE response and linked to that MDN article as well.